### PR TITLE
avoid sending modules in parent for dynamic imports

### PR DESCRIPTION
**Pull request by sokra**
### Description

Keeps track of available modules when importing new chunk groups. Omits assets that are already available in the parent chunk group(s) when determining assets to be included in chunks.

### Testing Instructions

Dynamic imports should not include modules that are already in the parent chunks

- [x] Auto label
fixes WEB-599

### CODE REVIEW

File: crates/next-core/src/next_client_component/with_client_chunks.rs [hunk 1]
```diff
@@ -154,20 +168,8 @@ impl ChunkItem for WithClientChunksChunkItem {
     }
 
     #[turbo_tasks::function]
-    async fn references(&self) -> Result<AssetReferencesVc> {
-        let inner = self.inner.await?;
-        Ok(AssetReferencesVc::cell(vec![
-            WithClientChunksAssetReference {
```
Comment (alexkirsz) [inline crates/next-core/src/next_client_component/with_client_chunks.rs:160]: Was this intended? The reference struct isn't used anywhere else.

File: crates/turbopack-core/src/chunk/available_assets.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+use turbo_tasks_hash::Xxh3Hash64Hasher;
+
+use super::{ChunkableAssetReference, ChunkableAssetReferenceVc, ChunkingType};
+use crate::{
+    asset::{Asset, AssetVc, AssetsSetVc},
+    reference::AssetReference,
+};
+
+#[turbo_tasks::value]
+pub struct AvailableAssets {
```
Comment (alexkirsz) [inline crates/turbopack-core/src/chunk/available_assets.rs:18]: Add a comment to explain what this is used for.

File: crates/turbopack-core/src/chunk/available_assets.rs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    #[turbo_tasks::function]
+    pub async fn hash(self) -> Result<U64Vc> {
+        let this = self.await?;
+        let mut hasher = Xxh3Hash64Hasher::new();
+        if let Some(parent) = this.parent {
+            hasher.write_value(parent.hash().await?);
+        }
+        for root in &this.roots {
+            hasher.write_value(root.ident().to_string().await?);
+        }
```
Comment (alexkirsz) [inline crates/turbopack-core/src/chunk/available_assets.rs:57]:
```suggestion
        if let Some(parent) = this.parent {
            hasher.write_value(1);
            hasher.write_value(parent.hash().await?);
        } else {
            hasher.write_value(0);
        }
        hasher.write_value(this.roots.len());
        for root in &this.roots {
            hasher.write_value(root.ident().to_string().await?);
        }
```

File: crates/turbopack-core/src/chunk/mod.rs [hunk 4]
```diff
@@ -242,6 +260,10 @@ pub enum ChunkingType {
     PlacedOrParallel,
     /// Asset is always placed in a separate chunk that is loaded in parallel.
     Parallel,
+    ///  Asset is always placed in a separate chunk that is loaded in parallel.
```
Comment (alexkirsz) [inline crates/turbopack-core/src/chunk/mod.rs:263]:
```suggestion
    /// Asset is always placed in a separate chunk that is loaded in parallel.
```

File: crates/turbopack-ecmascript/src/chunk/mod.rs [hunk 5]
```diff
@@ -355,6 +401,11 @@ impl Asset for EcmascriptChunk {
             modifiers.extend(evaluate.entry_modules_ids.iter().map(|id| id.to_string()));
         }
 
+        // Available assets are include
```
Comment (alexkirsz) [inline crates/turbopack-ecmascript/src/chunk/mod.rs:404]:
```suggestion
        // Available assets are included
```

File: crates/turbopack-ecmascript/src/references/pattern_mapping.rs [hunk 6]
```diff
@@ -157,8 +157,11 @@ impl PatternMappingVc {
 
         if let Some(chunkable) = ChunkableAssetVc::resolve_from(asset).await? {
             if *resolve_type == ResolveType::EsmAsync {
+                // Passing no available_assets and no current_availability_root works here
+                // because the manifest loader has an id that is independent of them. So luckily
+                // we don't need chunk dependenent code generation.
```
Comment (alexkirsz) [inline crates/turbopack-ecmascript/src/references/pattern_mapping.rs:162]:
```suggestion
                // we don't need chunk dependent code generation.
```

File: crates/next-core/src/next_client_chunks/with_chunks.rs [hunk 7]
```diff
@@ -49,16 +45,35 @@ impl Asset for WithChunksAsset {
...
 #[turbo_tasks::value_impl]
 impl ChunkableAsset for WithChunksAsset {
     #[turbo_tasks::function]
-    fn as_chunk(self_vc: WithChunksAssetVc, context: ChunkingContextVc) -> ChunkVc {
-        EcmascriptChunkVc::new(context, self_vc.as_ecmascript_chunk_placeable()).into()
+    fn as_chunk(
+        self_vc: WithChunksAssetVc,
+        context: ChunkingContextVc,
+        available_assets: Option<AvailableAssetsVc>,
+        current_availability_root: Option<AssetVc>,
```
Comment (alexkirsz) [inline crates/next-core/src/next_client_chunks/with_chunks.rs:68]: Feels like these two are almost always passed together. Can we make them part of a single `struct Availability`?

File: crates/next-core/src/next_client_component/with_client_chunks.rs (hunk 1 above)
Comment (sokra) [reply crates/next-core/src/next_client_component/with_client_chunks.rs:160]: yeah, it was weird before... Now it's simpler

Comment (alexkirsz) [reply crates/next-core/src/next_client_component/with_client_chunks.rs:160]: You should remove the `WithClientChunksAssetReference` then.

Comment (sokra) [reply crates/next-core/src/next_client_component/with_client_chunks.rs:164]: That's still use above

File: crates/turbopack-core/src/chunk/availablility_info.rs [hunk 8]
```diff
@@ -0,0 +1,40 @@
+use super::available_assets::AvailableAssetsVc;
+use crate::asset::AssetVc;
+
+#[turbo_tasks::value(serialization = "auto_for_input")]
+#[derive(PartialOrd, Ord, Hash, Clone, Copy, Debug)]
+pub enum AvailablilityInfo {
```
Comment (alexkirsz) [inline crates/turbopack-core/src/chunk/availablility_info.rs:6]:
```suggestion
pub enum AvailabilityInfo {
```

**Review by alexkirsz — APPROVED** [review]

File: crates/next-core/src/next_client_chunks/with_chunks.rs [hunk 9]
```diff
@@ -49,14 +49,10 @@ impl Asset for WithChunksAsset {
...
         Ok(AssetReferencesVc::cell(vec![
-            ChunkGroupReferenceVc::new(self_vc.chunk_group()).into(),
-            ChunkListReferenceVc::new(
-                this.server_root,
-                self_vc.chunk_group(),
-                self_vc.chunk_list_path(),
-            )
-            .into(),
+            ChunkGroupReferenceVc::new(chunk_group).into(),
+            ChunkListReferenceVc::new(this.server_root, chunk_group).into(),
```
Comment (alexkirsz) [inline crates/next-core/src/next_client_chunks/with_chunks.rs:55]: Could this be possibly part of the `ChunkGroupReference` transitive references?

File: crates/turbopack-core/src/chunk/dev.rs [hunk 10]
```diff
@@ -223,8 +223,14 @@ impl ChunkingContext for DevChunkingContext {
...
-    fn chunk_list_path(self_vc: DevChunkingContextVc, ident: AssetIdentVc) -> FileSystemPathVc {
-        self_vc.chunk_path(ident.with_modifier(chunk_list_modifier()), ".json")
+    fn chunk_list_path(
+        self_vc: DevChunkingContextVc,
+        entry_chunk_path: FileSystemPathVc,
+    ) -> FileSystemPathVc {
+        self_vc.chunk_path(
+            AssetIdentVc::from_path(entry_chunk_path).with_modifier(chunk_list_modifier()),
+            ".json",
+        )
```
Comment (alexkirsz) [inline crates/turbopack-core/src/chunk/dev.rs:233]: Getting `[UpdateServer]: error invalid uri character` when trying to subscribe to the chunk list updates with:

```
{"type":"subscribe","path":"_next/static/chunks/[root of the server]__next_static_chunks_.env_12715c.js_e3a1b5._.json","headers":null}
```

**Review by alexkirsz — CHANGES_REQUESTED** [review]

File: crates/turbopack-core/src/chunk/dev.rs (hunk 10 above)
Comment (sokra) [reply crates/turbopack-core/src/chunk/dev.rs]: fixed

**Review by alexkirsz — APPROVED** [review]

**Pull request merged**